display_list/display_list_path_effect.h

```diff
@@ -0,0 +1,319 @@
+
+  DlDiscretePathEffect(const DlDiscretePathEffect& discrete_effect)
+      : DlDiscretePathEffect(discrete_effect.segLength_,
+                             discrete_effect.dev_,
+                             discrete_effect.seedAssist_) {}
+  DlDiscretePathEffect(const DlDiscretePathEffect* discrete_effect)
+      : DlDiscretePathEffect(discrete_effect->segLength_,
+                             discrete_effect->dev_,
+                             discrete_effect->seedAssist_) {}
+
+  DlPathEffectType type() const override { return DlPathEffectType::kDiscrete; }
+  size_t size() const override { return sizeof(*this); }
+
+  static std::shared_ptr<DlDiscretePathEffect> Make(SkScalar segLength,
+                                                    SkScalar dev,
+                                                    uint32_t seedAssist = 0) {
+    return std::make_shared<DlDiscretePathEffect>(segLength, dev, seedAssist);
+  }
+
+  std::shared_ptr<DlPathEffect> shared() const override {
+    return std::make_shared<DlDiscretePathEffect>(this);
+  }
+
+  const DlDiscretePathEffect* asDiscrete() const override { return this; }
+
+  sk_sp<SkPathEffect> skia_object() const override {
+    return SkDiscretePathEffect::Make(segLength_, dev_, seedAssist_);
+  }
+
+ protected:
+  bool equals_(DlPathEffect const& other) const override {
+    FML_DCHECK(other.type() == DlPathEffectType::kDiscrete);
+    auto that = static_cast<DlDiscretePathEffect const*>(&other);
+    return segLength_ == that->segLength_ && dev_ == that->dev_ &&
+           seedAssist_ == that->seedAssist_;
+  }
+
+  SkScalar segLength_;
+  SkScalar dev_;
+  uint32_t seedAssist_ = 0;
+};
+
+class DlUnknownPathEffect final : public DlPathEffect {
+ public:
+  DlUnknownPathEffect(sk_sp<SkPathEffect> effect)
+      : sk_path_effect_(std::move(effect)) {}
+  DlUnknownPathEffect(const DlUnknownPathEffect& effect)
+      : DlUnknownPathEffect(effect.sk_path_effect_) {}
+  DlUnknownPathEffect(const DlUnknownPathEffect* effect)
+      : DlUnknownPathEffect(effect->sk_path_effect_) {}
+
+  DlPathEffectType type() const override { return DlPathEffectType::kUnknown; }
+  size_t size() const override { return sizeof(*this); }
+
+  std::shared_ptr<DlPathEffect> shared() const override {
+    return std::make_shared<DlUnknownPathEffect>(this);
+  }
+
+  sk_sp<SkPathEffect> skia_object() const override { return sk_path_effect_; }
+
+  virtual ~DlUnknownPathEffect() = default;
+
+ protected:
+  bool equals_(const DlPathEffect& other) const override {
+    FML_DCHECK(other.type() == DlPathEffectType::kUnknown);
+    auto that = static_cast<DlUnknownPathEffect const*>(&other);
+    return sk_path_effect_ == that->sk_path_effect_;
+  }
+
+ private:
+  sk_sp<SkPathEffect> sk_path_effect_;
+};
+class DlPairPathEffect : public DlPathEffect {
+ protected:
+  DlPairPathEffect(sk_sp<SkPathEffect> pe0, sk_sp<SkPathEffect> pe1)
+      : fPE0(std::move(pe0)), fPE1(std::move(pe1)) {}
```

Comment:
Keep in mind that any object that contains a shared_ptr cannot be Pod allocated into the DisplayList buffer. Note how some of the other types support both pod buffer allocation for the objects that are just a big chunk of raw data, Sk instantiation if they contain Sk data under the hood (primarily the DlUnknown variants), and Shared instantiation if they contain shared_ptr members - with an associated override of `equals()` in the `Op` structure.